### PR TITLE
[ext] feat: allow the users to save their preferences

### PR DESCRIPTION
**Pull request by GresilleSiffle**
**related to** #1560 

---

This PR adds the possibility to customize the recommendations languages used by the extension and the front end's recommendations page. The extension allows anonymous users to save their preferences locally, while authenticated users will have their preferences saved in the back end.

The strategy to determine which languages to use in the extension for anonymous users is the following:
- use the preferences saved locally if possible
- else fallback to the legacy mechanism used previously (local storage key shared between extension and front end)
- else use the navigator language if it's English or French
- else English

This strategy ensures that users without anything configured, and user with a navigator languages different from the "main" languages of Tournesol will always see the Tournesol recommendations on YouTube. The UI should be explicit enough to guide those users to preferences form.

The strategy to determine which languages to use in the extension for authenticated users is the same as the previous one, except that the first step is a call to the Tournesol API. 

**How to access the preferences from the extension?**

The browser actions menu, always accessible, now contains a Preferences button that open the preferences page. In the future we will move the Search on/off setting to the Preferences page.

capture

The Tournesol container provides also a direct link to the Preferences page.

capture

In addition to the previous methods, Firefox also displays a link to the Preferences page in its add-ons management page.  

capture

**to-do**

*page discoverability*
- [x] add a link to the preferences in the extension menu
- [x] add a link to the preferences in the Tournesol container

*ext refactor*
- [x] delete the old language mechanism

*ext recommendations*
- use the preferred languages when requesting the recommendations
  - [x] storage.local or browser language if in ['en', 'fr'] or 'en' for anonymous users
  - [x] API or or browser language if in ['en', 'fr'] or 'en' for authenticated users
- [x] if the preferred languages are not defined, try to load the legacy `recommendationsLanguages`

*ext options page - anonymous users*
- [x] init the languages with the legacy `recommendationsLanguages` key when there are no user settings
- [x] allow the selection of several languages
- [x] load their preferences from the extension storage
- [x] save their preferences in the extension storage
- [x] add a visual feedback after a successful save
- [x] add an invitation to create an account

*ext options page - authenticated users*
- [x] init the languages with the legacy `recommendationsLanguages` key when there are no user settings
- [x] allow the selection of several languages
- [x] load their preferences from the API
- [x] save their preferences with the API
- [x] add a visual feedback after a successful save
- [x] refresh the settings when the tab is open?
- [x] remove the sub menu when `embed=1?`

*front end refactor*
- [x] add the new settings `recommendations__default_languages`
- in the recommendations page, when the parameter `language` is not defined
  - [x] use the setting `recommendations__default_languages`
  - [x] fallback to the old local storage value for
    - [x] authenticated users without preferences
    - [x] anonymous users 
  - [x] fallback to languages of the navigator in case of need
- [x] fix the tests 
- [x] add e2e tests 

*future*

- transform the search enabled / disabled into a setting
- [x] in the extension's preferences, when the user is logged activate the "Tournesol Account" tab by default
- delete the content script reading the languages from the local storage? (this legacy mechanism is not required anymore)
- proposition: display the authenticated / anonymous status of the users in the Tournesol container (so that it's clear which kind of settings are used, local or authenticated)

#### Current

The users can easily edit their local preferences and their "Account" preferences at the same time.

capture

---

capture2

#### Firefox

capture

#### Chrome

captue

### CODE REVIEW

**Review by amatissart — COMMENTED** [review]
LGTM. My comments are very minor.

File: browser-extension/src/browserAction/menu.html [hunk 1]
```diff
@@ -72,5 +72,6 @@
     <button id="rate_later">Rate later</button>
     <button id="analysis">Analysis</button>
     <button id="search_state">Search is off</button>
+    <button id="preferences">Preferences</button>
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply browser-extension/src/browserAction/menu.html:75]: Exactly, I plan do it in the next pull requests.

There is a list of "to-do in the future" at the end of the PR description that contains the refactor of the search state.

File: browser-extension/src/options/options.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+ */
+
+/**
+ * Trigger custom behaviours after a specific tab is displayed.
+ */
+const afterTabDisplay = (tabName) => {
+  if (tabName === 'tournesol-account') {
+    const iframe = document.getElementById('iframe-tournesol-preferences');
+    iframe.setAttribute('src', iframe.getAttribute('src'));
+  }
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply browser-extension/src/options/options.js:155]: That's right it's a bit disturbing.

It's not really necessary, the iframe can work perfectly without the forced refresh. I added it to be sure the settings the user is seeing are up-to-date. This covers this supposedly unusual case:
- I open the extension preferences (the iframe is loaded)
- I update my preference directly from the Tournesol website
- I click on the Tournesol Account tab in the extension preferences
- the settings I'm seeing are now obsolete, if I click on save I'll erase my last modifications

I know this case might be very rare, and refreshing the iframe all the time to cover those few exceptions might feel overkill.

We can probably mitigate this case in other ways (such as providing a refresh button, and displaying a message to invite the users to refresh their preferences before submitting new ones).

File: browser-extension/src/utils.js [hunk 3]
```diff
@@ -150,3 +169,99 @@ export const isNavigatorLang = (lang) => {
...
+ * Tournesol API.
+ *
+ * Fallback to the extension local settings if the user is not authenticated,
+ * or if an error occured during the request.
+ *
+ * Return a list of ISO 639-1 language codes that can be used to fetch the
+ * recommendations from the Tournesol API.
+ */
+export const getRecommendationsLanguages = async (user) => {
+  if (user === 'authenticated') {
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply browser-extension/src/utils.js]: Fixed by https://github.com/tournesol-app/tournesol/pull/1683/commits/0852b2b666d137ed8c0f69ab725bd8a3f0cb731a

File: browser-extension/src/options/options.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply browser-extension/src/options/options.js]: Yes unfortunately :(

I think I found a more elegant solution for my problem: with https://github.com/tournesol-app/tournesol/pull/1683/commits/896a58c143eef3c1c1bdf11f1bbbb9b79efc9f0d the Tournesol Account tab will be open by default for authenticated users. This way, when the preferences page is open, the displayed settings will always be up-to-date. This also remove the blink effect caused by the forced refresh of the iframe.

**Comment by GresilleSiffle** [comment]
As suggested by @aidanjungo , I added two buttons allowing to easily select and deselect all languages.

capture

**Review by amatissart — APPROVED** [review]

**Pull request merged**